utilities/transactions/write_prepared_txn_db.h

```diff
@@ -110,8 +115,125 @@ class WritePreparedTxnDB : public PessimisticTransactionDB {
   virtual void ReleaseSnapshot(const Snapshot* snapshot) override;
 
   // Check whether the transaction that wrote the value with sequence number seq
-  // is visible to the snapshot with sequence number snapshot_seq
-  bool IsInSnapshot(uint64_t seq, uint64_t snapshot_seq) const;
+  // is visible to the snapshot with sequence number snapshot_seq.
+// Returns true if commit_seq <= snapshot_seq
+  inline bool IsInSnapshot(uint64_t prep_seq,
+                                      uint64_t snapshot_seq, uint64_t smallest_prepare = 0) const {
+  ROCKS_LOG_DETAILS(
+      info_log_, "IsInSnapshot %" PRIu64 " in %" PRIu64 " smalleest %" PRIu64,
+      prep_seq, snapshot_seq, smallest_prepare);
+  // Here we try to infer the return value without looking into prepare list.
+  // This would help avoiding synchronization over a shared map.
+  // TODO(myabandeh): optimize this. This sequence of checks must be correct but
+  // not necessary efficient
+  if (prep_seq == 0) {
+    // Compaction will output keys to bottom-level with sequence number 0 if
+    // it is visible to the earliest snapshot.
+    ROCKS_LOG_DETAILS(
+        info_log_, "IsInSnapshot %" PRIu64 " in %" PRIu64 " returns %" PRId32,
+        prep_seq, snapshot_seq, 1);
+    return true;
+  }
+  if (snapshot_seq < prep_seq) {
+    // snapshot_seq < prep_seq <= commit_seq => snapshot_seq < commit_seq
+    ROCKS_LOG_DETAILS(
+        info_log_, "IsInSnapshot %" PRIu64 " in %" PRIu64 " returns %" PRId32,
+        prep_seq, snapshot_seq, 0);
+    return false;
+  }
+  if (!delayed_prepared_empty_.load(std::memory_order_acquire)) {
+    // We should not normally reach here
+    ReadLock rl(&prepared_mutex_);
+    // TODO(myabandeh): also add a stat
+    ROCKS_LOG_WARN(info_log_, "prepared_mutex_ overhead %" PRIu64,
+                   static_cast<uint64_t>(delayed_prepared_.size()));
+    if (delayed_prepared_.find(prep_seq) != delayed_prepared_.end()) {
+      // Then it is not committed yet
+      ROCKS_LOG_DETAILS(
+          info_log_, "IsInSnapshot %" PRIu64 " in %" PRIu64 " returns %" PRId32,
+          prep_seq, snapshot_seq, 0);
+      return false;
+    }
+  }
+  if (prep_seq < smallest_prepare) {
```

Comment:
this is the only change but should be "after" checking delayed_prepared_. the reason is that after recovery delayed_prepared_ could have items in it while smallest_prepare did not capture them since it only looked at the PrepareHeap.

Reply:
Add inline comment for this? Also are we able to check delayed_prepared after recovery and update smallest_prepare accordingly?

Reply:
It is an interesting optimization. I leave it as a TODO since I already ran the benchmarks with the current implementation, and would need to do that all again should a major change happen.